doc/source/rllib-training.rst

```diff
@@ -520,6 +520,142 @@ Custom metrics can be accessed and visualized like any other training result:
 
 .. image:: custom_metric.png
 
+Customized Exploration Behavior (Training and Evaluation)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+RLlib offers a unified top-level API to configure and customize an agent’s
+exploration behavior, including the decisions (how and whether) to sample
+actions from distributions (stochastically or deterministically).
+The setup can be done via using built-in Exploration classes
+(see `this package <https://github.com/ray-project/ray/blob/master/rllib/utils/exploration/>`__),
+which are specified (and further configured) inside ``Trainer.config["exploration_config"]``.
+Besides using built-in classes, one can sub-class any of
+these built-ins, add custom behavior to it, and use that new class in
+the config instead.
+
+Every policy has-an instantiation of one of the Exploration (sub-)classes
+(described below). This Exploration object is created from the Trainer’s
+``config[“exploration_config”]`` dict, which specifies the class to use via the
+special “type” key, as well as all constructor arguments (via all other keys),
+e.g.:
+
+.. code-block:: python
+
+    # in Trainer.config:
+    "exploration_config": {
+        "type": "StochasticSampling",  # <- Special `type` key provides class information
+        "[c'tor arg]" : "[value]",  # <- Add any needed constructor args here.
+        # etc
+    }
+    # ...
+
+Every Exploration class implements the ``get_exploration_action`` method:
+
+.. code-block:: python
+
+    def get_exploration_action(self,
+                               distribution_inputs,
+                               action_dist_class,
+                               model=None,
+                               explore=True,
+                               timestep=None):
+        """Returns a (possibly) exploratory action and its log-likelihood.
+
+        Given the Model's logits outputs and action distribution, returns an
+        exploratory action.
+
+        Args:
+            distribution_inputs (any): The output coming from the model,
+                ready for parameterizing a distribution
+                (e.g. q-values or PG-logits).
+            action_dist_class (class): The action distribution class
+                to use.
+            model (ModelV2): The Model object.
+            explore (bool): True: "Normal" exploration behavior.
+                False: Suppress all exploratory behavior and return
+                    a deterministic action.
+            timestep (int): The current sampling time step. If None, the
+                component should try to use an internal counter, which it
+                then increments by 1. If provided, will set the internal
```

Comment:
Is this different from `timesteps_total` / `num_steps_sampled`?? We should make sure to use the same timestep counter / terminology everywhere to avoid confusion.

Reply:
The policy always passes `self.global_timestep` into the call to `get_exploration_action`. It's the same behavior across all Policy types. This prop is updated in `Policy.on_global_var_update -> global_vars["timestep"]`, so it should always be the total number of sample steps.

Reply:
I added that explanation to the docs as well.